core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -102,27 +110,52 @@ public void execute() throws Exception {
         }
 
         totalShards.incrementAndGet();
-        pendingShards.incrementAndGet();
+        pendingActions.incrementAndGet();
         primaryResult = primary.perform(request);
         final ReplicaRequest replicaRequest = primaryResult.replicaRequest();
         assert replicaRequest.primaryTerm() > 0 : "replicaRequest doesn't have a primary term";
         if (logger.isTraceEnabled()) {
             logger.trace("[{}] op [{}] completed on primary for request [{}]", primaryId, opType, request);
         }
 
-        performOnReplicas(primaryId, replicaRequest);
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        ClusterState clusterState = clusterStateSupplier.get();
+        final List<ShardRouting> shards = getShards(primaryId, clusterState);
+
+        if (markUnavailableShardsAsStale) {
+            Set<String> inSyncAllocationIds = getInSyncAllocationIds(primaryId, clusterState);
+
+            if (inSyncAllocationIds.isEmpty() == false && shards.isEmpty() == false) {
+                Set<String> availableAllocationIds = shards.stream()
+                    .map(ShardRouting::allocationId)
+                    .filter(Objects::nonNull)
+                    .map(AllocationId::getId)
+                    .collect(Collectors.toSet());
+
+                // if inSyncAllocationIds contains allocation ids of shards that don't exist in RoutingTable, mark copies as stale
+                for (String allocationId : Sets.difference(inSyncAllocationIds, availableAllocationIds)) {
+                    // mark copy as stale
+                    pendingActions.incrementAndGet();
+                    replicasProxy.markShardCopyAsStale(primaryId, allocationId, replicaRequest.primaryTerm(),
+                        ReplicationOperation.this::decPendingAndFinishIfNeeded,
+                        ReplicationOperation.this::onPrimaryDemoted,
+                        throwable -> decPendingAndFinishIfNeeded()
+                    );
+                }
+            }
+        }
+
+        performOnReplicas(replicaRequest, shards);
```

Comment:
@ywelsch I'm not sure if this is a possible issue. Imagine the following scenario: we have a two-node cluster with 1 primary shard and 1 replica.  Both are master eligible nodes.  The node holding the primary is also the master (lets call it `M`) and the other node we'll call `D`, which holds the replica shard.  Imagine `D` goes down.  When `D` goes down, we no longer remove the allocation id for the replica shard from the cluster state.  Now, a write request comes into `M`, and the write is applied to the primary shard on `M`.  Then we come to the replication phase and determine that the allocation ID residing on `D` needs to be marked as stale, so we send the request to fail the shard to `M`, which happens in a separate thread.  Meanwhile, we attempt to perform the replication operation for the write, and realize there are no active replicas to replicate to, so we finish up and acknowledge the write to the client.  After ACK'ing the write to the client, but before processing the cluster state update with the failed shard, `M` is shut down.  Now suppose we start up `D`.  Since its master eligible, it will become master and we don't have an updated cluster state update that removed the replica allocation ID, hence `D` promotes the replica to primary and we have lost an acknowledged write.  

I may well be missing something here, but this scenario came to my mind and I wanted to validate it.